GCS_MAVLink: Add FLIGHT_INFORMATION message

nexton-winjeel opened the pull request:
Adds ~~a new `FLIGHT_DURATION`~~ the `FLIGHT_INFORMATION` message to report the time that the vehicle armed and took off (if known).

This message is needed if connecting a GCS to a vehicle that is already armed/flying, as otherwise these times can't be accurately determined.

Also adds an accessor to `AP_Arming` to record the arming time.

~~Requires: https://github.com/ArduPilot/mavlink/pull/336~~

IamPete1 commented:
I wonder if the new metrics would be useful as part of `AP_Stats`, you can get flight time from there, but currently you have to pull the param value.

peterbarker commented:
@nexton-winjeel I think this PR is now no-longer-required as we're going to use `FLIGHT_INFORMATION` instead, correct?

nexton-winjeel commented:
@peterbarker: It's not required as-is, but I'll convert it to use `FLIGHT_INFORMATION` instead of the now non-existent `FLIGHT_DURATION`. Will get to it early next week.

peterbarker commented:
> @peterbarker: It's not required as-is, but I'll convert it to use `FLIGHT_INFORMATION` instead of the now non-existent `FLIGHT_DURATION`. Will get to it early next week.

Long week?

nexton-winjeel commented:
> Long week?

Very. Still not my oldest open PR though...

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -5960,6 +5960,51 @@ void GCS_MAVLINK::send_autopilot_state_for_gimbal_device() const
+                }
+                break;
+            }
+            default: {
```
```suggestion
            MAV_LANDED_STATE_UNDEFINED:
            MAV_LANDED_STATE_ON_GROUND:
```

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -5960,6 +5960,51 @@ void GCS_MAVLINK::send_autopilot_state_for_gimbal_device() const
+        flight_info.last_landed_state = current_landed_state;
+    }
+
+    // We send time since boot (in micros) rather than a UTC timestamp, as this
+    // works better with SITL when SIM_SPEEDUP > 1
```
This comment is out of date once the patches to the mavlink message upstream are brought back.

Instead it should read something like the fields are misnamed...

Also, the flag value is a little problematic with the problematic `ARMING_REQUIRE` flag on Rover and Plane.  There's always an argument that they're not really armed in that state, but just pretend to be.... but still, how does this message work for an `ARMING_REQUIRE` 0 vehicle?

nexton-winjeel replied:
> [...] There's always an argument that they're not really armed in that state, but just pretend to be...

They will set the `MAV_MODE_FLAG_SAFETY_ARMED` flag of the `base_mode` field of the `HEARTBEAT` message, so as far as a GCS knows, they _are_ armed.

> [...] how does this message work for an `ARMING_REQUIRE` 0 vehicle?

It will report arming time as `0` if `AP_Arming::arm()` is never called. This field shouldn't be treated as a proxy for "is armed", it needs to be read in conjunction with the arming sate reported in the `HEARTBEAT`.

peterbarker reviewed (changes requested):
Would be nice to get the message updates down from mavlink/mavlink.

I don't think we need to tie this with AP_Stats now.  It's not a huge amount of shared state.

peterbarker commented on libraries/AP_Arming/AP_Arming.h:
```diff
@@ -100,6 +100,9 @@ class AP_Arming {
     bool is_armed_and_safety_off() const;
 
+    // Returns the time since boot (in millis) that we armed. Returns 0 if disarmed.
+    uint32_t arm_time_ms() const { return is_armed() ? last_arm_time_ms : 0; }
```
64-bit

tridge replied:
should be 64 bit so it doesn't wrap at 49 days

tridge commented on libraries/GCS_MAVLink/GCS.h:
```diff
@@ -402,6 +402,15 @@ class GCS_MAVLINK
+#if AP_MAVLINK_MSG_FLIGHT_INFORMATION_ENABLED
+    struct {
+        MAV_LANDED_STATE last_landed_state;
+        uint32_t takeoff_time_ms;
```
64 bit

peterbarker commented:
Closes https://github.com/ArduPilot/ardupilot/issues/25284

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -334,7 +334,7 @@ void GCS_MAVLINK::send_battery_status(const uint8_t instance) const
     float current, consumed_mah, consumed_wh;
     const int8_t percentage = battery_remaining_pct(instance);
-    
+
```
Please remove the whitespace changes.

Fix the whitespace problems as you change the code.

nexton-winjeel replied:
Done.

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -5964,6 +5964,58 @@ void GCS_MAVLINK::send_autopilot_state_for_gimbal_device() const
+#if AP_MAVLINK_MSG_FLIGHT_INFORMATION_ENABLED
+void GCS_MAVLINK::send_flight_information()
+{
+    const uint32_t time_boot_us = AP_HAL::millis64();
```
```suggestion
    const uint64_t time_boot_us = AP_HAL::millis64();
```

nexton-winjeel replied:
Fixed.

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -5964,6 +5964,58 @@ void GCS_MAVLINK::send_autopilot_state_for_gimbal_device() const
+        switch (current_landed_state) {
+            case MAV_LANDED_STATE_IN_AIR:
+            case MAV_LANDED_STATE_TAKEOFF:
+            case MAV_LANDED_STATE_LANDING: {
```
Don't really need to open braces if you're not declaring variables.

Doesn't need to be fixed, just maybe leave them out in future

nexton-winjeel replied:
Defensive programming habit. Fixed while I was in there...

peterbarker commented on libraries/GCS_MAVLink/GCS_config.h:
```diff
@@ -135,3 +135,7 @@
 #endif
+
+#ifndef AP_MAVLINK_MSG_FLIGHT_INFORMATION_ENABLED
+#define AP_MAVLINK_MSG_FLIGHT_INFORMATION_ENABLED 1
```
What's the flash size cost of this new feature?

Will it cause us to start overflowing on smaller boards?

Should we add it into `minimize_common.inc`? (@andyp1per and @Hwurzburg might have ideas on that)

We don't need to change this as part of this PR.

nexton-winjeel replied:
> What's the flash size cost of this new feature?

~728 bytes.

```
          Target          Text (B)  Data (B)  BSS (B)  Total Flash Used (B)  Free Flash (B)  External Flash Used (B)
          ----------------------------------------------------------------------------------------------------------
Disabled: bin/arducopter   1840268      3920   258372               1844188          121876  Not Applicable 
Enabled:  bin/arducopter   1840996      3920   258372               1844916          121148  Not Applicable  
```

peterbarker reviewed (changes requested):
Need to fix that type, but LGTM otherwise.

Hwurzburg reviewed (changes requested):
I think this should be disabled by def and a build option...not going to be used by 95% of users....or at least disabled in minimize include...also needs file for build server and features list added

peterbarker commented:
> I think this should be disabled by def and a build option...not going to be used by 95% of users....or at least disabled in minimize include...also needs file for build server and features list added

This is going to be a good/useful message for a lot of people.  I plan on patching MAVProxy for it to fix the resetting-flight-time problem it has.

@nexton-winjeel please add an entry into `minimize_common.inc` so this isn't built on the tiny boards.  Also entries to `extract_features.py` and `build_options.py` so users can turn the feature on/off as desired.

nexton-winjeel commented:
> @nexton-winjeel please add an entry into `minimize_common.inc` so this isn't built on the tiny boards. Also entries to `extract_features.py` and `build_options.py` so users can turn the feature on/off as desired.

Done.

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -5964,6 +5964,55 @@ void GCS_MAVLINK::send_autopilot_state_for_gimbal_device() const
+
+    mavlink_msg_flight_information_send(
+        chan,
+        time_boot_us,
```
This is actually a `time_boot_ms` field...

nexton-winjeel replied:
Fixed.

peterbarker reviewed (changes requested):
We need a mapping in `GCS_MAVLINK::mavlink_id_to_ap_message_id` for this message

Also a fix for ms vs us

Consider grabbing my patch from my `pr/Winjeel/upstream/add-FLIGHT_DURATION-message` branch

peterbarker commented:
In-progress MAVProxy support is here: https://github.com/ArduPilot/MAVProxy/pull/1438

Works, but I've broken other stuff in a refactor

peterbarker commented:
Ping @nexton-winjeel - might still be viable for 4.6 - but even than had better be soon :-)

nexton-winjeel commented:
@peterbarker: Can I get you to review this again?

patrickelectric commented:
Can we have this merged ?

peterbarker commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -6074,6 +6077,56 @@ void GCS_MAVLINK::send_autopilot_state_for_gimbal_device() const
+#if AP_MAVLINK_MSG_FLIGHT_INFORMATION_ENABLED
+void GCS_MAVLINK::send_flight_information()
+{
+    const uint64_t time_boot_micros = AP_HAL::micros64();
+    const uint32_t time_boot_ms = static_cast<uint32_t>(time_boot_micros / 1000);
+
+    // This field is misnamed as `arming_time_utc` in MAVLink. However, it is
+    // not a UTC time, it is the microseconds since boot.
+    const uint64_t arm_time_us = AP::arming().arm_time_us();
```
Prefer to declare/define variables as late as possible.

nexton-winjeel replied:
In this case, they're declared in the order they are used in the `mavlink_msg_flight_information_send` function.

peterbarker commented:
I can't force-push this branch.

I've rebased this branch over here: https://github.com/ArduPilot/ardupilot/pull/29541 - just resolved the conflicts.

I intend to merge that other PR once it passes CI.  And then the MAVProxy one.

peterbarker commented:
> @peterbarker: Can I get you to review this again?

Apologies - missed this request.

nexton-winjeel commented:
> I can't force-push this branch.
> 
> I've rebased this branch over here: #29541 - just resolved the conflicts.
> 
> I intend to merge that other PR once it passes CI. And then the MAVProxy one.

I've rebased this one if you want to merge this instead of the other branch.

peterbarker commented:
Merged, thanks!
